### ISSUE TITLE
Vim9: "b:match_words" causes "%"'s matchit to conflate "func" variable type with ":func" command

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#7734]
**Describe the bug**

In Vim9 script, if we use the default matchit plugin, we cannot press `%` on `:enddef` to jump to the matching `:def`, if the function contains a variable declaration whose type is `func`.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9
        packadd matchit
        filetype plugin on
        var lines =<< trim END
            def Func()
                var Ref: func
            enddef
        END
        setline(1, lines)
        set ft=vim
        norm! G
        norm %
    EOF
    )

The cursor is on `func`.

**Expected behavior**

The cursor is on `def`.

**Environment**

 - Vim version: 8.2 Included patches: 1-2394
 - OS: Ubuntu 16.04.7 LTS
 - Terminal: xterm(363)

**Additional context**

The issue doesn't come from the matchit plugin.  It comes from `b:match_words` which is set in the default Vim filetype plugin, and uses a too permissive pattern to match `:function`.  I think it should ignore `func` if it's preceded by a colon and whitespace.

This patch should fix the issue:
```diff
diff --git a/runtime/ftplugin/vim.vim b/runtime/ftplugin/vim.vim
index 6adadbda9..c1a4a8544 100644
--- a/runtime/ftplugin/vim.vim
+++ b/runtime/ftplugin/vim.vim
@@ -89,7 +89,7 @@ endif
 if exists("loaded_matchit")
   let b:match_ignorecase = 0
   let b:match_words =
-	\ '\<\%(fu\%[nction]\|def\)\>)\@!:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
+        \ '\<\%(\%(\:\s\+\)\@<!fu\%[nction]\|def\)\>)\@!:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
  	\ '\<\(wh\%[ile]\|for\)\>:\<brea\%[k]\>:\<con\%[tinue]\>:\<end\(w\%[hile]\|fo\%[r]\)\>,' .
 	\ '\<if\>:\<el\%[seif]\>:\<en\%[dif]\>,' .
 	\ '{:},' .
```
---

I found this while using a custom mapping which gives the number of lines of code in a Vim function.  It wrongly reports 0 lines inside a `:def` function if it contains a variable declaration whose type is `func`.  That's because it relies on the `%` command of [`vim-matchup`](https://github.com/andymass/vim-matchup), which – like matchit – relies on `b:match_words`.

**Comment by brammool** [comment]
> In Vim9 script, if we use the default matchit plugin, we cannot press `%` on `:enddef` to jump to the matching `:def`, if the function contains a variable declaration whose type is `func`.
> 
> **To Reproduce**
> 
> Run this shell command:
> 
>     vim -Nu NONE -S <(cat <<'EOF'
>         vim9
>         packadd matchit
>         filetype plugin on
>         var lines =<< trim END
>             def Func()
>                 var Ref: func
>             enddef
>         END
>         setline(1, lines)
>         set ft=vim
>         norm! G
>         norm %
>     EOF
>     )
> 
> The cursor is on `func`.
> 
> **Expected behavior**
> 
> The cursor is on `def`.
> 
> **Environment**
> 
>  - Vim version: 8.2 Included patches: 1-2394
>  - OS: Ubuntu 16.04.7 LTS
>  - Terminal: xterm(363)
> 
> **Additional context**
> 
> The issue doesn't come from the matchit plugin.  It comes from the default Vim filetype plugin which uses a too permissive pattern to match `:function`.  I think it should ignore `func` if it's preceded by a colon and whitespace.
> 
> This patch should fix the issue:
> ```diff
> diff --git a/runtime/ftplugin/vim.vim b/runtime/ftplugin/vim.vim
> index 6adadbda9..c1a4a8544 100644
> --- a/runtime/ftplugin/vim.vim
> +++ b/runtime/ftplugin/vim.vim
> @@ -89,7 +89,7 @@ endif
>  if exists("loaded_matchit")
>    let b:match_ignorecase = 0
>    let b:match_words =
> -	\ '\<\%(fu\%[nction]\|def\)\>)\@!:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
> +        \ '\<\%(\%(\:\s\+\)\@<!fu\%[nction]\|def\)\>)\@!:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
>   	\ '\<\(wh\%[ile]\|for\)\>:\<brea\%[k]\>:\<con\%[tinue]\>:\<end\(w\%[hile]\|fo\%[r]\)\>,' .
>  	\ '\<if\>:\<el\%[seif]\>:\<en\%[dif]\>,' .
>  	\ '{:},' .
> ```

That doesn't really work, one can use ":function", the colon is ignored.

There are other occurences of "function" that are not followed by a
matching "endfunction", such as "function name".  Looks like the best
way is to check for a following name and then a parenthesis.

  let b:match_words =
	\ '\<\%(fu\%[nction]\|def\)\s\+\S\+(:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
 	\ '\<\(wh\%[ile]\|for\)\>:\<brea\%[k]\>:\<con\%[tinue]\>:\<end\(w\%[hile]\|fo\%[r]\)\>,' .
	\ '\<if\>:\<el\%[seif]\>:\<en\%[dif]\>,' .
	\ '{:},' .
	\ '\<try\>:\<cat\%[ch]\>:\<fina\%[lly]\>:\<endt\%[ry]\>,' .
	\ '\<aug\%[roup]\s\+\%(END\>\)\@!\S:\<aug\%[roup]\s\+END\>,'

-- 
hundred-and-one symptoms of being an internet addict:
218. Your spouse hands you a gift wrapped magnet with your PC's name
     on it and you accuse him or her of genocide.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
> let b:match_words =
   >       \ '\<\%(fu\%[nction]\|def\)\s\+\S\+(:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
   >       ...

There could be whitespace between the function name and the parenthesis:

    \ '\<\%(fu\%[nction]\|def\)\s\+\S\+\s*(:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
                                       ^^^

Although, I would consider that as a syntax error.  But it doesn't raise any error in Vim script legacy, nor in Vim9 script:
```vim
vim9
def Func   ()
    echo 'test'
enddef
Func()
```
Considering that Vim9 is very strict regarding whitespace, I think this is a bug; which looks very similar to https://github.com/vim/vim/issues/6496; the latter is about whitespace between a function name and the parenthesis in a call (not in a definition).

**Comment by lacygoill** [comment]
Fixed by https://github.com/vim/vim/commit/e0e3917554327f2524066f89fbbef9c83c1535da .

**Comment by lacygoill** [comment]
> Fixed by e0e3917 .

But the patch has introduced a regression.  Now, `%` fails to jump between `:function` and `:endfunction`, if `:function` is followed by a bang.  I think this patch should fix the issue:
```diff
diff --git a/runtime/ftplugin/vim.vim b/runtime/ftplugin/vim.vim
index d39ad9579..d5081918c 100644
--- a/runtime/ftplugin/vim.vim
+++ b/runtime/ftplugin/vim.vim
@@ -94,7 +94,7 @@ if exists("loaded_matchit")
   "   func name
   " require a parenthesis following, then there can be an "endfunc".
   let b:match_words =
-	\ '\<\%(fu\%[nction]\|def\)\s\+\S\+(:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
+	\ '\<\%(fu\%[nction]\|def\)!\=\s\+\S\+(:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
  	\ '\<\(wh\%[ile]\|for\)\>:\<brea\%[k]\>:\<con\%[tinue]\>:\<end\(w\%[hile]\|fo\%[r]\)\>,' .
 	\ '\<if\>:\<el\%[seif]\>:\<en\%[dif]\>,' .
 	\ '{:},' .
```

**Comment by brammool** [comment]
> But the patch has introduced a regression.  Now, `%` fails to jump
> between `:function` and `:endfunction`, if `:function` is followed by
> a bang.  I think this patch should fix the issue:
> ```diff
> diff --git a/runtime/ftplugin/vim.vim b/runtime/ftplugin/vim.vim
> index d39ad9579..d5081918c 100644
> --- a/runtime/ftplugin/vim.vim
> +++ b/runtime/ftplugin/vim.vim
> @@ -94,7 +94,7 @@ if exists("loaded_matchit")
>    "   func name
>    " require a parenthesis following, then there can be an "endfunc".
>    let b:match_words =
> -	\ '\<\%(fu\%[nction]\|def\)\s\+\S\+(:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
> +	\ '\<\%(fu\%[nction]\|def\)!\=\s\+\S\+(:\<retu\%[rn]\>:\<\%(endf\%[unction]\|enddef\)\>,' .
>   	\ '\<\(wh\%[ile]\|for\)\>:\<brea\%[k]\>:\<con\%[tinue]\>:\<end\(w\%[hile]\|fo\%[r]\)\>,' .
>  	\ '\<if\>:\<el\%[seif]\>:\<en\%[dif]\>,' .
>  	\ '{:},' .
> ```

Thanks, looks like this works.

-- 
"A mouse can be just as dangerous as a bullet or a bomb."
             (US Representative Lamar Smith, R-Texas)

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
